File: tests/anonymize.test.js
import { describe, expect, test } from "bun:test";
import { spawnSync } from "node:child_process";
import { readdir, readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { anonymize, deanonymize, redact, guessAnonymizationCategory } from "../src/js/fake/anon.js";
import { preload } from "../src/js/fake/data.js";

await preload();

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const dataDir = path.join(__dirname, "data");

async function loadFixtures() {
	const entries = await readdir(dataDir);
	const fixtures = [];
	for (const entry of entries.filter((name) => name.endsWith(".json")).sort()) {
		const payload = JSON.parse(await readFile(path.join(dataDir, entry), "utf8"));
		fixtures.push([entry, payload]);
	}
	return fixtures;
}

function getPath(value, pathValue) {
	return pathValue.split(".").reduce((current, part) => {
		if (Array.isArray(current)) return current[Number(part)];
		return current[part];
	}, value);
}

describe("anonymize fixtures", async () => {
	const fixtures = await loadFixtures();

	test("loads all eight shared fixtures", () => {
		expect(fixtures).toHaveLength(8);
	});

	for (const [name, fixture] of fixtures) {
		test(`${name}: preserves symbol-like values`, async () => {
			const result = await anonymize(fixture.document, 42);
			for (const pathValue of fixture.expectations.symbol_paths) {
				expect(getPath(result.value, pathValue)).toEqual(getPath(fixture.document, pathValue));
			}
		});

		test(`${name}: roundtrip restores the original`, async () => {
			const result = await anonymize(fixture.document, 42);
			const restored = await deanonymize(result.value, 42, 0.25, null, null, null, result.mapping);
			expect(restored).toEqual(fixture.document);
		});

		test(`${name}: mapping has expected shape`, async () => {
			const result = await anonymize(fixture.document, 42);
			expect(result.mapping.version).toBe(3);
			expect(result.mapping.rules).toEqual({ variance: 0.25 });
			expect(result.mapping.seed_fingerprint).toHaveLength(16);
			expect(result.mapping.values).toBeObject();
		});

		test(`${name}: anonymize is deterministic`, async () => {
			const first = await anonymize(fixture.document, 42);
			const second = await anonymize(fixture.document, 42);
			expect(first).toEqual(second);
		});

		test(`${name}: PII fields are anonymized consistently`, async () => {
			const result = name === "secrets.json"
				? await redact(fixture.document, 42)
				: await anonymize(fixture.document, 42);
			for (const pathValue of fixture.expectations.pii_paths) {
				expect(getPath(result.value, pathValue)).not.toEqual(getPath(fixture.document, pathValue));
			}
		});

		test(`${name}: Python and JavaScript outputs are interchangeable`, async () => {
			const jsResult = await anonymize(fixture.document, 42);
			const python = spawnSync(
				"python3",
				[
					"-c",
					[
						"import json, pathlib, sys",
						"sys.path.insert(0, 'src/py')",
						"import fake",
						"fixture = json.load(open(sys.argv[1], encoding='utf-8'))",
						"print(json.dumps(fake.anonymize(fixture['document'], seed=42)))",
					].join("; "),
					path.join(dataDir, name),
				],
				{ cwd: path.join(__dirname, ".."), encoding: "utf8" },
			);
			expect(python.status).toBe(0);
			expect(jsResult).toEqual(JSON.parse(python.stdout));
		});
	}

	test("dates are fuzzed within weeks but not misclassified as phones", async () => {
		const fixture = fixtures.find(([name]) => name === "insurance-policy.json")[1];
		const result = await anonymize(fixture.document, 42);
		expect(result.value.policy.holder.dateOfBirth).not.toBe(fixture.document.policy.holder.dateOfBirth);
		expect(result.value.policy.claimSummary.incidentDate).not.toBe(fixture.document.policy.claimSummary.incidentDate);
		expect(result.value.policy.holder.dateOfBirth).toMatch(/^\d{4}-\d{2}-\d{2}$/);
	});

	test("full-name fields are anonymized conservatively", async () => {
		const payload = {
			preferredName: "Isaac Watts",
			PersonName: "Joanne Watts",
			adviserName: "Ellen Drenon",
			fileName: "Isaac Watts.pdf",
			companyName: "Watts Lending",
			$type: "com.example.PersonName",
		};
		const result = await anonymize(payload, 42);
		expect(result.value.preferredName).not.toBe(payload.preferredName);
		expect(result.value.PersonName).not.toBe(payload.PersonName);
		expect(result.value.adviserName).not.toBe(payload.adviserName);
		expect(result.value.fileName).toBe(payload.fileName);
		expect(result.value.companyName).toBe(payload.companyName);
		expect(result.value.$type).toBe(payload.$type);
	});

	test("hints and pii_string fallback", async () => {
		const payload = {
			NameOfPerson: "WeirdValue123",
			lastName: "Smith",
			someDate: "2020-05-05",
			emailField: "not-an-email",
		};
		const result = await anonymize(payload, 42, 0.25, { NameOfPerson: "pii_string", lastName: "last_name" });
		expect(result.value.NameOfPerson).not.toBe(payload.NameOfPerson);
		expect(result.value.NameOfPerson).toMatch(/^\[PII-/);
		expect(result.value.lastName).not.toBe(payload.lastName);
		expect(result.value.someDate).not.toBe(payload.someDate);
		expect(result.value.emailField).not.toBe(payload.emailField);
  		expect(result.value.emailField).toContain("@");

		// numeric values under email/address/role/history keys must be preserved as-is
		const payload2 = {
			OrganisationGroupAnnualReviewEmailDays: 5,
			PersonAddressHistory: 0,
			AdviserPublicCNARole: "User",
			AdviserDisputeResolutionService: "FDRS",
		};
		const result2 = await anonymize(payload2, 42);
		expect(result2.value.OrganisationGroupAnnualReviewEmailDays).toBe(5);
		expect(result2.value.PersonAddressHistory).toBe(0);
		expect(result2.value.AdviserPublicCNARole).toBe("User");
		expect(result2.value.AdviserDisputeResolutionService).toBe("FDRS");
  	});

 	test("guess, whitelist, blacklist, datetime time jitter", async () => {
 		expect(await guessAnonymizationCategory("PersonLastName", "Miller")).toBe("last_name");
 		expect(await guessAnonymizationCategory("PersonTwoFactorAuthStatus", "Disabled")).toBeNull();
 		expect(await guessAnonymizationCategory("FooStatus", "Bar", { whitelist: { FooStatus: "symbol" } })).toBe("symbol");
 		expect(await guessAnonymizationCategory("SecretField", "val", { blacklist: ["SecretField"] })).toBeNull();

		// numeric and role enums should not be miscategorized
		expect(await guessAnonymizationCategory("OrganisationGroupAnnualReviewEmailDays", 5)).toBeNull();
		expect(await guessAnonymizationCategory("PersonAddressHistory", 0)).toBeNull();
		expect(await guessAnonymizationCategory("AdviserPublicCNARole", "User")).toBeNull();
		// acronym/enum id under service key -> symbol (preserved, not name)
		expect(await guessAnonymizationCategory("AdviserDisputeResolutionService", "FDRS")).toBe("symbol");
  		const p = { created: "2026-04-08T03:12:25.715309Z" };
  		const r = await anonymize(p, 42);
  		expect(r.value.created).not.toBe(p.created);
  		// whitelist preserve
  		const r2 = await anonymize(p, 42, 0.25, null, { created: "symbol" });
  		expect(r2.value.created).toBe(p.created);
  	});

  	test("phone structure is preserved", async () => {
  		const cases = [
  			"232-123-1234",
  			"+1 (321)-1231-12311",
  			"88-81-12-13",
  			"+44 20 7946 0958",
  			"+1 (415) 555-0141",
  			"1234567",
  			"123456789012345",
  		];
  		for (const c of cases) {
  			const res = (await anonymize({ p: c }, 42)).value.p;
  			expect(res).not.toBe(c);
  			expect(res.replace(/\D/g, "").length).toBe(c.replace(/\D/g, "").length);
  			expect(res.replace(/\d/g, "D")).toBe(c.replace(/\d/g, "D"));
  			const mapping = (await anonymize({ p: c }, 42)).mapping;
  			const restored = (await deanonymize({ p: res }, 42, 0.25, null, null, null, mapping)).p;
  			expect(restored).toBe(c);
  		}
  	});
  });

File: tests/anonymize_test.py
import json
import pathlib
import re
import subprocess
import sys
import unittest


ROOT = pathlib.Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "src" / "py"))

import fake  # noqa: E402


DATA_DIR = ROOT / "tests" / "data"


def load_fixtures():
	fixtures = []
	for path in sorted(DATA_DIR.glob("*.json")):
		with path.open(encoding="utf-8") as handle:
			fixtures.append((path.name, json.load(handle)))
	return fixtures


def get_path(value, path):
	current = value
	for part in path.split("."):
		if isinstance(current, list):
			current = current[int(part)]
		else:
			current = current[part]
	return current


class AnonymizeFeatureTest(unittest.TestCase):
	maxDiff = None

	def test_fixture_inventory(self):
		fixtures = load_fixtures()
		self.assertEqual(8, len(fixtures))
		for name, fixture in fixtures:
			with self.subTest(name=name):
				self.assertIn("document", fixture)
				self.assertIn("expectations", fixture)
				self.assertIn("symbol_paths", fixture["expectations"])
				self.assertIn("pii_paths", fixture["expectations"])

	def test_symbol_fields_are_preserved(self):
		for name, fixture in load_fixtures():
			original = fixture["document"]
			result = fake.anonymize(original, seed=42)
			anonymized = result["value"]
			for path in fixture["expectations"]["symbol_paths"]:
				with self.subTest(name=name, path=path):
					self.assertEqual(get_path(original, path), get_path(anonymized, path))

	def test_dates_are_fuzzed_but_not_misclassified_as_phones(self):
		fixture = json.loads((DATA_DIR / "insurance-policy.json").read_text(encoding="utf-8"))
		original = fixture["document"]
		result = fake.anonymize(original, seed=42)["value"]
		# all date-keyed date values are now fuzzed within ~3 weeks
		self.assertNotEqual(
			original["policy"]["holder"]["dateOfBirth"],
			result["policy"]["holder"]["dateOfBirth"],
		)
		self.assertNotEqual(
			original["policy"]["claimSummary"]["incidentDate"],
			result["policy"]["claimSummary"]["incidentDate"],
		)
		# still not turned into phones (guard preserved)
		self.assertTrue(result["policy"]["holder"]["dateOfBirth"].startswith("19") or result["policy"]["holder"]["dateOfBirth"].startswith("20"))

	def test_roundtrip_restores_original_document(self):
		for name, fixture in load_fixtures():
			original = fixture["document"]
			result = fake.anonymize(original, seed=42)
			restored = fake.deanonymize(
				result["value"], seed=42, mapping=result["mapping"]
			)
			with self.subTest(name=name):
				self.assertEqual(original, restored)

	def test_mapping_has_expected_shape(self):
		for name, fixture in load_fixtures():
			redact = name == "secrets.json"
			mapping = fake.anonymize(fixture["document"], seed=42, redact_secrets=redact)["mapping"]
			with self.subTest(name=name):
				self.assertEqual(3, mapping["version"])
				self.assertEqual({"variance": 0.25}, mapping["rules"])
				self.assertEqual(16, len(mapping["seed_fingerprint"]))
				self.assertIsInstance(mapping["values"], dict)
				if name == "secrets.json":
					self.assertTrue(mapping.get("secrets", False))

	def test_anonymize_is_deterministic(self):
		for name, fixture in load_fixtures():
			first = fake.anonymize(fixture["document"], seed=42)
			second = fake.anonymize(fixture["document"], seed=42)
			with self.subTest(name=name):
				self.assertEqual(first, second)

	def test_phone_structure_is_preserved(self):
		cases = [
			"232-123-1234",
			"+1 (321)-1231-12311",
			"88-81-12-13",
			"+44 20 7946 0958",
			"+1 (415) 555-0141",
			"1234567",
			"123456789012345",
		]
		for c in cases:
			with self.subTest(value=c):
				res = fake.anonymize({"p": c}, seed=42)["value"]["p"]
				self.assertNotEqual(res, c)
				self.assertEqual(len(re.sub(r"\D", "", res)), len(re.sub(r"\D", "", c)))
				self.assertEqual(re.sub(r"\d", "D", res), re.sub(r"\d", "D", c))
				restored = fake.deanonymize({"p": res}, seed=42, mapping=fake.anonymize({"p": c}, seed=42)["mapping"])["p"]
				self.assertEqual(restored, c)

	def test_pii_fields_are_anonymized(self):
		for name, fixture in load_fixtures():
			original = fixture["document"]
			result = fake.anonymize(original, seed=42, redact_secrets=True)
			anonymized = result["value"]
			for path in fixture["expectations"]["pii_paths"]:
				with self.subTest(name=name, path=path):
					self.assertNotEqual(get_path(original, path), get_path(anonymized, path))

	def test_full_name_fields_are_anonymized_conservatively(self):
		payload = {
			"preferredName": "Isaac Watts",
			"PersonName": "Joanne Watts",
			"adviserName": "Ellen Drenon",
			"fileName": "Isaac Watts.pdf",
			"companyName": "Watts Lending",
			"$type": "com.example.PersonName",
		}
		result = fake.anonymize(payload, seed=42)["value"]
		self.assertNotEqual(payload["preferredName"], result["preferredName"])
		self.assertNotEqual(payload["PersonName"], result["PersonName"])
		self.assertNotEqual(payload["adviserName"], result["adviserName"])
		self.assertEqual(payload["fileName"], result["fileName"])
		self.assertEqual(payload["companyName"], result["companyName"])
		self.assertEqual(payload["$type"], result["$type"])

	def test_hints_and_pii_string_fallback(self):
		payload = {
			"NameOfPerson": "WeirdValue123",
			"lastName": "Smith",
			"someDate": "2020-05-05",
			"emailField": "not-an-email",
		}
		result = fake.anonymize(payload, seed=42, hints={"NameOfPerson": "pii_string", "lastName": "last_name"})["value"]
		self.assertNotEqual(payload["NameOfPerson"], result["NameOfPerson"])
		self.assertTrue(result["NameOfPerson"].startswith("[PII-"))
		self.assertNotEqual(payload["lastName"], result["lastName"])
		# date key without hint still fuzzed if matches date pattern
		self.assertNotEqual(payload["someDate"], result["someDate"])
		# email context forces email transform
		self.assertNotEqual(payload["emailField"], result["emailField"])
		self.assertIn("@", result["emailField"])

		# numeric values under email/address/role/history keys must be preserved as-is (not miscategorized)
		payload2 = {
			"OrganisationGroupAnnualReviewEmailDays": 5,
			"PersonAddressHistory": 0,
			"AdviserPublicCNARole": "User",
			"AdviserDisputeResolutionService": "FDRS",
		}
		result2 = fake.anonymize(payload2, seed=42)["value"]
		self.assertEqual(result2["OrganisationGroupAnnualReviewEmailDays"], 5)
		self.assertEqual(result2["PersonAddressHistory"], 0)
		self.assertEqual(result2["AdviserPublicCNARole"], "User")
		self.assertEqual(result2["AdviserDisputeResolutionService"], "FDRS")


	def test_guess_whitelist_blacklist_and_datetime_time(self):
		# guess func
		self.assertEqual(fake.guessAnonymizationCategory("PersonLastName", "Miller"), "last_name")
		self.assertEqual(fake.guessAnonymizationCategory("PersonTwoFactorAuthStatus", "Disabled"), None)
		# whitelist forces even for status like
		self.assertEqual(fake.guessAnonymizationCategory("FooStatus", "Bar", whitelist={"FooStatus": "symbol"}), "symbol")
		# blacklist preserves
		self.assertIsNone(fake.guessAnonymizationCategory("SecretField", "val", blacklist=["SecretField"]))

		# numeric and role enums should not be miscategorized
		self.assertIsNone(fake.guessAnonymizationCategory("OrganisationGroupAnnualReviewEmailDays", 5))
		self.assertIsNone(fake.guessAnonymizationCategory("PersonAddressHistory", 0))
		self.assertIsNone(fake.guessAnonymizationCategory("AdviserPublicCNARole", "User"))
		# acronym/enum id under service key -> symbol (preserved, not name)
		self.assertEqual(fake.guessAnonymizationCategory("AdviserDisputeResolutionService", "FDRS"), "symbol")
		# datetime with time should jitter both
		p = {"created": "2026-04-08T03:12:25.715309Z"}
		r = fake.anonymize(p, seed=42)["value"]["created"]
		self.assertNotEqual(p["created"], r)
		# date part may differ, time part should often differ
		self.assertTrue(r.startswith("2026-") or r.startswith("2025-") or r.startswith("2027-"))  # within jitter
		# use whitelist to preserve a category
		r2 = fake.anonymize(p, seed=42, whitelist={"created": "symbol"})["value"]["created"]
		self.assertEqual(p["created"], r2)

	def test_python_and_js_outputs_are_interchangeable(self):
		for name, fixture in load_fixtures():
			python_result = fake.anonymize(fixture["document"], seed=42)
			fixture_path = DATA_DIR / name
			script = """
import { anonymize } from './src/js/fake/anon.js';
import { readFileSync } from 'node:fs';
const fixture = JSON.parse(readFileSync(process.argv[process.argv.length - 1], 'utf8'));
const result = await anonymize(fixture.document, 42);
console.log(JSON.stringify(result));
""".strip()
			completed = subprocess.run(
				["bun", "--input-type=module", "-e", script, str(fixture_path)],
				check=True,
				capture_output=True,
				text=True,
				cwd=str(ROOT),
			)
			js_result = json.loads(completed.stdout)
			with self.subTest(name=name):
				self.assertEqual(python_result, js_result)


if __name__ == "__main__":
	unittest.main()
